python/EMR_SECURITY_GROUPS_RESTRICTED/EMR_SECURITY_GROUPS_RESTRICTED.py

```diff
@@ -116,36 +115,52 @@ def get_sg_cluster_mapping(cluster_list, emr_client):
                 sg_list_combined.add(security_group)
     return sg_cluster_map, sg_list_combined
 
-def get_open_security_groups(sg_list, ec2_client):
+def get_open_security_groups(sg_list, config_client):
     open_sg_list = set([])
-    paginator = ec2_client.get_paginator('describe_security_groups')
-    page_iterator = paginator.paginate(GroupIds=list(sg_list))
-    for page in page_iterator:
-        for security_group in page['SecurityGroups']:
-            if is_sg_open_ipv4(security_group):
-                open_sg_list.add(security_group['GroupId'])
-            elif is_sg_open_ipv6(security_group):
-                open_sg_list.add(security_group['GroupId'])
+    resource_keys = []
+    for security_group in sg_list:
+        resource_key = {
+            "resourceType": "AWS::EC2::SecurityGroup",
+            "resourceId": security_group
+        }
+        resource_keys.append(resource_key)
+    while resource_keys:
+        configuration_items = get_config_items(config_client, resource_keys[:100])
+        del resource_keys[:100]
+    for item in configuration_items:
+        if is_sg_open_ipv4(json.loads(item["configuration"])):
+            open_sg_list.add(json.loads(item["configuration"])['groupId'])
+        elif is_sg_open_ipv6(json.loads(item["configuration"])):
+            open_sg_list.add(json.loads(item["configuration"])['groupId'])
     return open_sg_list
 
 def is_sg_open_ipv4(sg_description):
-    for rule in sg_description['IpPermissions']:
-        for ip_range in rule['IpRanges']:
-            if ip_range['CidrIp'] == '0.0.0.0/0':
+    for rule in sg_description['ipPermissions']:
+        for ip_range in rule['ipv4Ranges']:
+            if ip_range['cidrIp'] == '0.0.0.0/0':
                 return True
     return False
 
 def is_sg_open_ipv6(sg_description):
-    for rule in sg_description['IpPermissions']:
-        for ip_range in rule['Ipv6Ranges']:
-            if ip_range['CidrIpv6'] == '::/0':
+    for rule in sg_description['ipPermissions']:
+        for ip_range in rule['ipv6Ranges']:
+            if ip_range['cidrIpv6'] == '::/0':
                 return True
     return False
 
 def evaluate_parameters(rule_parameters):
     valid_rule_parameters = rule_parameters
     return valid_rule_parameters
 
+def get_config_items(config_client, resource_keys):
+    res = config_client.batch_get_resource_config(resourceKeys=resource_keys)
+    configuration_items = res["baseConfigurationItems"]
+    while res["unprocessedResourceKeys"]:
```

Comment:
You are risking an infinite loop here: if the cluster get created, then the rule run before the sg is registered in Config. The sg will remain into the said 'unprocessedResourceKeys' for up to 15 minutes (or more).
I suggest: after 3 tries -> mark the sg as "unregistered", and do not post any information about the EMR cluster(s) containing it.

Reply:
That's just the thing! If the SG doesn't exist, it doesn't turn up in the unprocessed items list. It just seems to be skipped. The same happens if you provide some random bogus resource id. 
For example:
```
aws configservice batch-get-resource-config --resource-keys resourceType=AWS::EC2::SecurityGroup,resourceId=sg-0ff787b5ade90a1243
{
    "baseConfigurationItems": [],
    "unprocessedResourceKeys": []
}
```

Reply:
Oooh, I see now your point. Well there is the easy way, and we run with what is today. Or we need to compare resource_keys with the configuration_item. I think let's start with what we have first. We can move into CR.